emma-language/src/main/scala/org/emmalanguage/compiler/lang/FunSupport.scala

```diff
@@ -0,0 +1,478 @@
+                // compute type bindings sequence
+                val typesSeq = for {
+                  (tp, ta) <- tparams zip targs
+                } yield tp -> ta.typeSymbol.asType
+                // compute type bindings map
+                val typesMap = (for {
+                  (tp, ta) <- typesSeq
+                } yield tp.toType -> ta.toType).toMap.withDefault(t => t)
+
+                // compute term bindings sequence
+                val termsSeq = for {
+                  s <- (bndDefs diff parsUsedOnce).toSeq
+                } yield s -> Sym.copy(s)(
+                  name = TermName.fresh(s),
+                  tpe = get.tpe(s).map(typesMap),
+                  flags = if (parsUsedMany(s)) u.NoFlags else get.flags(s)
+                ).asTerm
+                // compute term bindings sequence
+                val termsMap = termsSeq.toMap
+
+                // compute a sequence of `symbol -> tree` substitutions
+                val substSeq = for {
+                  (ParDef(p, _, _), a) <- paramss.flatten zip argss.flatten
+                } yield {
+                  if (parsUsedOnce(p)) p -> a
+                  else p -> Ref(termsMap(p))
+                }
+
+                // compute prefix of ValDefs derived from ParDefs used more than once
+                val prefxSeq = for {
+                  (ParDef(p, _, _), a) <- paramss.flatten zip argss.flatten
+                  if parsUsedMany(p)
+                } yield ValDef(termsMap(p), a)
+
+                // substitute terms and types and inline arguments in the body
+                val substBody = ({
+                  Tree.rename(termsSeq ++ typesSeq: _*)
+                } andThen {
+                  Tree.subst(substSeq: _*)
+                }) (body)
+
+                // prepend ValDefs prefix to the substitution result
+                val prefxBody =
+                  if (prefxSeq.isEmpty) substBody
+                  else substBody match {
+                    case Block(stats, expr) => Block(prefxSeq ++ stats: _*)(expr)
+                    case expr => Block(prefxSeq: _*)(expr)
+                  }
+
+                val anyEq = u.typeOf[Any].member(TermName("==")).asMethod
+                val fixEqBody = BottomUp.transform({
+                  case DefCall(Some(lhs), `anyEq`, Seq(), Seq(rhs)) =>
+                    val specializedEq = Type.of(lhs).member(anyEq.name).asTerm
+                    DefCall(Some(lhs))(specializedEq)(Seq(rhs))
+                })(prefxBody).tree
+
+                Owner.at(owner)(fixEqBody)
+            })(snippet.tree).tree
+
+          result
+      }
+
+      CG.Snippet(result)
+    }
+
+    private lazy val calleAttrs: u.Tree => (Set[u.TermSymbol], Map[u.TermSymbol, Int]) =
+      BottomUp.withParUses.withBindDefs.traverseAny andThen {
+        case Attr.syn(_, bndDefs :: parUses :: _) => (bndDefs.keySet, parUses)
+      }
+
+    lazy val callGraph: u.Tree => CG = tree => {
```

Comment:
Discussed offline, and the conclusion was that since the definition of `f` doesn't get inlined into `g` in the first example, this can't cause an actual infinite inlining.